### PR TITLE
Enhances the buildmode tool

### PR DESCRIPTION
**Pull request by Crazylemon64**
Based a good part off of -tg-station's buildmode layout. From tgstation/-tg-station#15353, mostly. Credit to @AnturK!

Thing's I've added to this system, apart from tg's original refactor:
* Area mode reticules
* Made a new class of datum for buildmode, so that it doesn't need to use call() to interact with it
* Exits out of any click_intercepts when you enter a new mob, so you don't get stuck in buildmode without a UI

- [ ] Make an area fill mode
- [ ] Make an area edit mode, as in map areas

:cl:Crazylemon
rscadd: The buildmode area tool now shows reticules of what you've selected.
bugfix: Switching mobs while using the buildmode tool no longer screws up your UI.
tweak: Refactors buildmode so it's no longer a special-case system.
rscadd: Adds a copy mode to build mode, which lets you duplicate objects.
tweak: Any mob in buildmode can work at the fastest possible rate.
/:cl:

### CODE REVIEW

File: paradise.dme [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #define DEBUG
 // END_PREFERENCES
 // BEGIN_INCLUDE
+#include "stddef.dm"
```
Comment (tigercat2000) [inline paradise.dme]: it won't compile due to this

Comment (Crazylemon64) [inline paradise.dme]: I thought I left that out, dangit...

File: code/_onclick/ai.dm [hunk 2]
```diff
@@ -10,9 +10,8 @@
 	Note that AI have no need for the adjacency proc, and so this proc is a lot cleaner.
 */
 /mob/living/silicon/ai/DblClickOn(var/atom/A, params)
-	if(client.buildmode) // comes after object.Click to allow buildmode gui objects to be clicked
-		build_click(src, client.buildmode, params, A)
-		return
+	if(client.click_intercept)
+		client.click_intercept.InterceptClickOn(src, params, A)
```
Comment (Crazylemon64) [inline code/_onclick/ai.dm]: Oh shoot I forgot the returns

**Comment by Crazylemon64** [comment]
I've removed the limiters on speed for buildmode - there shouldn't be any problem with this, right?

**Comment by Regen1** [comment]
ENHANCE!

**Comment by Dumbdumn5** [comment]
We had an AREA MODE? ALL HAIL!

**Comment by Crazylemon64** [comment]
Well, it currently just uses a procedural generator for the area - there's probably nicer things we can do with it, like actually changing place's areas...

**Comment by Dumbdumn5** [comment]
Eh? How about an actual area mode, like making a box of walls or changing all the tiles in a selected area into plating? Not trying to push for it, but I was under the impression nobody wanted to touch buildmode, so I sort of saw it as a pipedream that anyone would.

**Comment by Crazylemon64** [comment]
Oh! That! Yes, that sounds good! I think I will add that, yes! But not immediately! Now is for sleeping!

**Comment by monster860** [comment]
Procedural generator for the area? Isn't that basically glorified station blueprints then?

**Comment by Crazylemon64** [comment]
No, no, it uses code from the `procedural_mapping` module to fill in the region with tiles. It doesn't actually modify areas like a blueprint would; That's next on the list.

**Comment by Dumbdumn5** [comment]
Can the copy mechanic copy ANYTHING or just objs?

**Comment by Crazylemon64** [comment]
@Dumbdumn5 It copies atoms, and transfers over any variables that aren't references to datums (as those could possibly break). This DOES mean quite a few things will break if you copy them, so you might want to be careful, when doing so...

**Comment by Tastyfish** [comment]
But it might work for doors, for example.

**Comment by tigercat2000** [comment]
@Crazylemon64 You might want to just copy the `appearance` var; It makes up all of the stuff that controls the apperance of an atom, and generally that's what you would want with a copy

**Comment by Crazylemon64** [comment]
@tigercat2000 But then it wouldn't let me copy, say, a wand of fireballs with -1 charges - if I wanted to just place down a bunch of a certain item, I could just use the advanced build mode for that

**Comment by tigercat2000** [comment]
@Crazylemon64 i suppose so

**Comment by Crazylemon64** [comment]
This is ready as-is, I no longer plan to add more to this. Merge if you want.

**Pull request merged**